Enforce SCRIPT_VERIFY_P2SH and SCRIPT_VERIFY_WITNESS from genesis

sdaftuar opened the pull request:
As discussed at the IRC meeting back in October (https://botbot.me/freenode/bitcoin-core-dev/2017-10-12/?msg=92231929&page=2), I had looked into the feasibility of enforcing P2SH and SCRIPT_VERIFY_WITNESS back to the genesis block.

The P2SH change is pretty straightforward -- there was only one historical block on mainnet that violated the rule, so I carved out an exception to it, similar to the way we have exceptions for the BIP30 violators.

The segwit change is not entirely as clear.  The code changes themselves are relatively straightforward: we can just always turn on SCRIPT_VERIFY_WITNESS whenever P2SH is active.  However conceptually, this amounts to splitting up BIP141 into two parts, the part that implements new script rules, and the part that handles witness commitments in blocks.  

Arguably though the script rules are really defined in BIP 143 anyway, and so this really amounts to backdating BIP 143 -- script rules for v0 segwit outputs -- back to genesis.  So maybe conceptually this isn't so bad...

I don't feel strongly about this change in either direction; I started working on it because I was searching for a way to simplify the way we understand and implement the consensus rules around segwit, but I'm not yet sure whether I think this achieves anything toward that goal.

ping @TheBlueMatt

sdaftuar committed:
Use P2SH consensus rules for all blocks

This commit moves P2SH activation back to the genesis block, with
a hardcoded exception for the one historical block in the chain that
violated this rule.

sdaftuar committed:
Separate NULLDUMMY enforcement from SEGWIT enforcement

This is in preparation for enforcing SCRIPT_VERIFY_WITNESS from
the genesis block.

sdaftuar committed:
Always enforce SCRIPT_VERIFY_WITNESS with P2SH

sdaftuar committed:
[qa] Remove some pre-activation segwit tests

This is in preparation for always enforcing SCRIPT_VERIFY_WITNESS.

sdaftuar commented:
Forgot to flag this: there is one mainnet block that violated P2SH, and one testnet block as well.  For now I carved out both, but perhaps we could consider migration to a new testnet (4?) instead and drop the testnet exception?  Seems like it might be cleaner to have a new testnet that always has P2SH and segwit from genesis....  Does anyone think this would be worth proposing?

MarcoFalke commented on src/validation.cpp:
```diff
@@ -1611,7 +1611,18 @@ static ThresholdConditionCache warningcache[VERSIONBITS_NUM_BITS];
+    unsigned int flags = fStrictPayToScriptHash ? SCRIPT_VERIFY_P2SH : SCRIPT_VERIFY_NONE;
 
     // Start enforcing P2SH (BIP16)
     if (pindex->nHeight >= consensusparams.BIP16Height) {
```
fixup nit: in commit 3a76e3c962516cdf7d577aa3dcbe0251b94e5355: You probably meant to use ```if (fStrictPayToScriptHash) { flags |= SCRIPT_VERIFY_P2SH; }```, as BIP16Height was removed and thus fails to compile.

TheBlueMatt commented:
Concept ACK. At a code level not so interesting, I suppose (except for maybe the P2SH backdating), but it's generally a nice property for us to, over time, enforce soft forks like SegWit universally.

On November 20, 2017 12:32:05 PM PST, Suhas Daftuar <notifications@github.com> wrote:
>As discussed at the IRC meeting back in October
>(https://botbot.me/freenode/bitcoin-core-dev/2017-10-12/?msg=92231929&page=2),
>I had looked into the feasibility of enforcing P2SH and
>SCRIPT_VERIFY_WITNESS back to the genesis block.
>
>The P2SH change is pretty straightforward -- there was only one
>historical block on mainnet that violated the rule, so I carved out an
>exception to it, similar to the way we have exceptions for the BIP30
>violators.
>
>The segwit change is not entirely as clear.  The code changes
>themselves are relatively straightforward: we can just always turn on
>SCRIPT_VERIFY_WITNESS whenever P2SH is active.  However conceptually,
>this amounts to splitting up BIP141 into two parts, the part that
>implements new script rules, and the part that handles witness
>commitments in blocks.  
>
>Arguably though the script rules are really defined in BIP 143 anyway,
>and so this really amounts to backdating BIP 143 -- script rules for v0
>segwit outputs -- back to genesis.  So maybe conceptually this isn't so
>bad...
>
>I don't feel strongly about this change in either direction; I started
>working on it because I was searching for a way to simplify the way we
>understand and implement the consensus rules around segwit, but I'm not
>yet sure whether I think this achieves anything toward that goal.
>
>ping @TheBlueMatt 
>You can view, comment on, or merge this pull request online at:
>
>  https://github.com/bitcoin/bitcoin/pull/11739
>
>-- Commit Summary --
>
>  * Use P2SH consensus rules for all blocks
>  * Separate NULLDUMMY enforcement from SEGWIT enforcement
>  * Always enforce SCRIPT_VERIFY_WITNESS with P2SH
>  * [qa] Remove some pre-activation segwit tests
> * scripted-diff: rename IsWitnessEnabled -> IsWitnessCommitmentEnabled
>
>-- File Changes --
>
>    M src/chainparams.cpp (3)
>    M src/consensus/params.h (2)
>    M src/miner.cpp (2)
>    M src/net_processing.cpp (8)
>    M src/validation.cpp (49)
>    M src/validation.h (5)
>    M src/wallet/rpcwallet.cpp (2)
>    M test/functional/p2p-segwit.py (2)
>    M test/functional/segwit.py (8)
>
>-- Patch Links --
>
>https://github.com/bitcoin/bitcoin/pull/11739.patch
>https://github.com/bitcoin/bitcoin/pull/11739.diff
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/bitcoin/bitcoin/pull/11739

sipa commented:
The code changes here don't look as bad as I would have expected. I'd prefer to encapsulate the exception block hashes in chainparams, though.

gmaxwell commented:
Concept ACK!   as far as testnet, restarting with a new testnet would be fine, but there are a bunch of very interesting test cases in testnet we should consider extracting. One possibility would be to instrument the codebase to call the gprof things to record branch coverage data at each block; an then make a list of testnet blocks that increased cover.  Seems like a real pain though.

sdaftuar commented:
Nits addressed.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -1608,16 +1608,28 @@ class WarningBitsConditionChecker : public AbstractThresholdConditionChecker
+    // BIP16 didn't become active until Apr 1 2012
+    // However, only one historical block violated the P2SH rules, so for simplicity, always leave P2SH
+    // on except for the one violating block.
+    if (!(pindex->phashBlock != nullptr && (*(pindex->phashBlock) == consensusparams.BIP16Exception))) {
```
Why not assert pindex->phashBlock? No properly-constructed CBlockIndex should have that unset, no?

sdaftuar replied:
It's unset when invoked via TestBlockValidity() (eg for mining).

TheBlueMatt replied:
Ah, that sounds like a bug in TBV, but that's for a separate PR, I suppose.

jtimon replied:
Yes, please, I always found checks to pindex->phashBlock == nullptr ugly and expensive to maintain documented (perhaps we can repeat the "pindex->phashBlock can be null if called by CreateNewBlock/TestBlockValidity" comment from the beginning and unified with the comment for fEnforceBIP30, to more easily delete them all later when they're not needed). There's the asset you want below, just needs to be moved to the top once the special cases are dealt with more explicitly.
But that's for a separate PR.

TheBlueMatt reviewed (commented):
utACK ad3e97a109541bb644dc1cb5e264a2954926bcac

laanwj commented:
Concept ACK

ryanofsky commented on src/validation.cpp:
```diff
@@ -1613,8 +1613,10 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consens
+    // BIP16 didn't become active until Apr 1 2012
+    // However, only one historical block violated the P2SH rules, so for simplicity, always leave P2SH
+    // on except for the one violating block.
+    if (!(pindex->phashBlock != nullptr && (*(pindex->phashBlock) == consensusparams.BIP16Exception))) {
```
In commit "Use P2SH consensus rules for all blocks"

Maybe check BIP16Exception.IsNull() to be a little more strict. Could also remove unnecessary nesting and negatives:

```c++
if (consensusparams.BIP16Exception.IsNull() ||
    pindex->phashBlock == nullptr ||
    *pindex->phashBlock != consensusparams.BIP16Exception)
```

ryanofsky commented on test/functional/p2p-segwit.py:
```diff
@@ -1905,7 +1905,7 @@ def run_test(self):
         self.test_witness_tx_relay_before_segwit_activation()
         self.test_block_relay(segwit_activated=False)
-        self.test_p2sh_witness(segwit_activated=False)
+        #self.test_p2sh_witness(segwit_activated=False)
```
In commit "[qa] Remove some pre-activation segwit tests"

Should actually remove

ryanofsky commented on test/functional/segwit.py:
```diff
@@ -189,17 +189,13 @@ def run_test(self):
-        self.log.info("Verify unsigned bare witness txs in versionbits-setting blocks are valid before the fork")
-        self.success_mine(self.nodes[2], wit_ids[NODE_2][WIT_V0][1], False) #block 428
-        self.success_mine(self.nodes[2], wit_ids[NODE_2][WIT_V1][1], False) #block 429
+        # TODO: verify spending a segwit v0 output is invalid before the fork
```
In commit "[qa] Remove some pre-activation segwit tests"

Note: TODO

jnewbery replied:
I don't understand this comment. What does 'before the fork' mean if segwit is active since genesis?

ryanofsky reviewed (commented):
utACK ad3e97a109541bb644dc1cb5e264a2954926bcac

gmaxwell commented:
Tested ACK

sdaftuar committed:
[qa] Test that v0 segwit outputs can't be spent pre-activation

Also updates the comments for an existing test, that now should be rewritten.
Includes changes suggested by John Newbery.

sdaftuar commented:
Addressed @ryanofsky's comments.  However I think I will need to rebase in order to for the new test I wrote to work...

jtimon reviewed (commented):
This needs a BIP IMO.
Concept ACK. if the segwit case is not clear yet, perhaps it should be left for later (in fact, I would prefer the 2 things in 2 different PRs either way, preferrably one on top of the other [perhaps open one with only p2sh and rebase this one of top of the new one], but all between the parenthesis is just PR bikeshedding).

utACK individual commits df7fad0a0b0f02ab2193673bf16ebffc5318778d d32bcd91c7748feb559848bf4db6da8fe4715494

sdaftuar commented:
Rebased and fixed the last commit to work with latest master.  Previous version of these commits is at https://github.com/sdaftuar/bitcoin/commits/11739.1

sdaftuar commented:
@jtimon Agreed -- I sent an [email to the bitcoin-dev](https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2018-January/015588.html) list recently with a link to the [draft BIP text](https://github.com/sdaftuar/bips/blob/p2sh-v0segwit-from-genesis/bip-sdaftuar-p2sh-v0segwit-from-genesis.mediawiki).

jnewbery commented on src/net_processing.cpp:
```diff
@@ -857,7 +857,7 @@ void PeerLogicValidation::NewPoWValidBlock(const CBlockIndex *pindex, const std:
     nHighestFastAnnounce = pindex->nHeight;
 
-    bool fWitnessEnabled = IsWitnessEnabled(pindex->pprev, Params().GetConsensus());
+    bool fWitnessEnabled = IsWitnessCommitmentEnabled(pindex->pprev, Params().GetConsensus());
```
Could also change the name of the local variable here:

`fWitnessEnabled` -> `witness_commitment_enabled`

jnewbery commented on src/chainparams.cpp:
```diff
@@ -75,9 +75,9 @@ class CMainParams : public CChainParams {
-        consensus.BIP16Height = 173805; // 00000000000000ce80a7e057163a4db1d5ad7b20fb6f598c9597b9665c8fb0d4 - April 1, 2012
         consensus.BIP34Height = 227931;
         consensus.BIP34Hash = uint256S("0x000000000000024b89b42a942fe0d9fea3bb44ab7bd1b19115dd6a759c0808b8");
+        consensus.BIP16Exception = uint256S("0x00000000000002dc756eebf4f49723ed8d30cc28a5f108eb94b1ba88ac4f9c22");
```
nit: prefer the BIP constants in numeric order (ie BIP 16 above BIP 34). Same for params.h below.

jnewbery commented on test/functional/p2p-segwit.py:
```diff
@@ -271,6 +271,50 @@ def test_unnecessary_witness_before_segwit_activation(self):
+
+        assert(len(self.utxo))
+
+        # Create two outputs, a p2wsh and p2sh-p2wsh
```
comment says p2sh-p2wsh, but the second output below is just a p2sh as far as I can tell.

jnewbery replied:
strange. The P2SH output can't be spent either. If I change the `test_witness_block()` call to a `test_transaction_acceptance()` call below, I see the following log:

`(mandatory-script-verify-flag-failed (Operation not valid with the current stack size) (code 16))`

sdaftuar replied:
Fixed -- I forgot to stick the p2sh redeem script in the scriptsig (and also inadvertently made this a p2sh, rather than p2sh-p2wsh).

jnewbery commented:
utACK up to 4e2b90d21690059dc496a17d2aee9fe11ce3a8b2.

I don't think the final commit adds any value and can be removed from this PR.

sdaftuar commented:
Github thought this needed to be rebased, so I did, though there were no conflicts to be resolved.
Original unsquashed version is https://github.com/sdaftuar/bitcoin/commits/11739.2.

jnewbery commented on test/functional/feature_segwit.py:
```diff
@@ -150,17 +150,13 @@ def run_test(self):
-        self.log.info("Verify unsigned bare witness txs in versionbits-setting blocks are valid before the fork")
-        self.success_mine(self.nodes[2], wit_ids[NODE_2][WIT_V0][1], False) #block 428
-        self.success_mine(self.nodes[2], wit_ids[NODE_2][WIT_V1][1], False) #block 429
+        # TODO: verify spending a segwit v0 output is invalid before the fork
```
I wouldn't bother adding this comment, which you're just going to remove in a later commit. I think you can also remove the comment above.

sdaftuar commented:
I decided to remove the scripted-diff commit, because in retrospect I think the existing function name, `IsWitnessEnabled`, is fine.

Addressed nits from @jnewbery and @ryanofsky and squashed.  Old version is: https://github.com/sdaftuar/bitcoin/commits/11739.3

sdaftuar commented:
Updated the last commit (which introduces a new test) with some fixes and cleanups suggested by @jnewbery.  Prior version of this PR is [11739.4](https://github.com/sdaftuar/bitcoin/commits/11739.4)

sdaftuar commented:
Pushed two more commits to fix a comment pointed out by @jnewbery, and fix a test bug that travis triggered.  Travis is passing now.  Squashed [11739.5](https://github.com/sdaftuar/bitcoin/commits/11739.5) -> 9080e76f3e5a2afb6352aa79763e6d4d976d7cce

jnewbery commented:
Tested ACK 9080e76f3e5a2afb6352aa79763e6d4d976d7cce.

Test changes look good. It'd be nice to do some general tidy-up to the segwit functional tests (`p2p_segwit.py` and `feature_segwit.py`), but we should leave that for another PR after #11398 is merged.

Sjors commented on src/chainparams.cpp:
```diff
@@ -75,7 +75,7 @@ class CMainParams : public CChainParams {
         strNetworkID = "main";
         consensus.nSubsidyHalvingInterval = 210000;
-        consensus.BIP16Height = 173805; // 00000000000000ce80a7e057163a4db1d5ad7b20fb6f598c9597b9665c8fb0d4 - April 1, 2012
+        consensus.BIP16Exception = uint256S("0x00000000000002dc756eebf4f49723ed8d30cc28a5f108eb94b1ba88ac4f9c22");
```
Nit: if we want to reuse this pattern: `consensus.BIP16Exceptions = [...]`?

Sjors commented on src/chainparams.cpp:
```diff
@@ -186,7 +186,7 @@ class CTestNetParams : public CChainParams {
         strNetworkID = "test";
         consensus.nSubsidyHalvingInterval = 210000;
-        consensus.BIP16Height = 514; // 00000000040b4e986385315e14bee30ad876d8b47f748025b26683116d21aa65
+        consensus.BIP16Exception = uint256S("0x00000000dd30457c001f4095d208cc1296b0eed002427aa599874af7a432b105");
```
Put a TODO here to remove it when the next testnet comes?

Sjors reviewed (commented):
Concept ACK. This sort of thing is also useful for sidechain-like projects where you want to track the bitcoin core codebase with ideally zero patches.

Sjors commented:
Should we also remove the relevant soft fork entries from `getblockchaininfo`?

jnewbery commented:
> Should we also remove the relevant soft fork entries from getblockchaininfo?

BIP 16 already is not shown in `getblockchaininfo` (it was a flag day upgrade rather than IsSuperMajority deployment). This PR doesn't move the segwit softfork height, it just enforces the `SCRIPT_VERIFY_WITNESS` rules to genesis.

instagibbs commented on src/validation.cpp:
```diff
@@ -1740,6 +1745,11 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consens
         flags |= SCRIPT_VERIFY_P2SH;
     }
 
+    // Enforce WITNESS rules whenever P2SH is in effect.
```
"Enforce WITNESS rules whenever P2SH is in effect and witness deployment is not defined."

... or something

sdaftuar replied:
I've improved the comments in latest version.

instagibbs reviewed (approved):
I am unable to comment on the changes here as I don't understand that section of code: https://github.com/bitcoin/bitcoin/pull/11739/commits/8b975bc6c6f3fd3889ba3f907d9e4a663525d240#diff-24efdb00bfbe56b140fb006b562cc70bR4034

I have also not tested the exact hashes given.

utACK otherwise

instagibbs commented on src/validation.cpp:
```diff
@@ -1730,8 +1730,13 @@ static unsigned int GetBlockScriptFlags(const CBlockIndex* pindex, const Consens
 
-    // Start enforcing P2SH (BIP16)
-    if (pindex->nHeight >= consensusparams.BIP16Height) {
+    // BIP16 didn't become active until Apr 1 2012
```
For the benefit of code readers, maybe say that as well :)

sdaftuar commented:
Updated comments to address nit from @instagibbs, and squashed (previous version is https://github.com/sdaftuar/bitcoin/commits/11739.6 for comparison).

sdaftuar commented:
Updated again to improve code comment around historical p2sh activation.  (Compare: https://github.com/sdaftuar/bitcoin/commits/11739.7)

sdaftuar commented:
Needed rebase.

jnewbery commented:
Tested ACK 8b56fc0b91eb4876004bdc92a7015b12b34a04ed

Travis failure was unrelated p2p_compactblocks.py flakiness. I've restared the job.

It'd be great to get this (and then #12360) in before v0.17.

sdaftuar commented:
Tests are passing now.

TheBlueMatt commented:
non-test-changes-re-utACK 8b56fc0b91eb4876004bdc92a7015b12b34a04ed (though I miss the s/IsWitnessEnabled/IsWitnessCommitmentEnabled/ scripted-diff change).

MarcoFalke commented:
utACK 8b56fc0b91eb4876004bdc92a7015b12b34a04ed

MarcoFalke commented:
This seemed ready for merge. Tough, would be nice to see one or two more re-ACKs, since this went through a rebase.
